pymeasure/instruments/thyracont/transmitter.py

```diff
@@ -0,0 +1,491 @@
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+#
+
+from enum import IntEnum
+
+from pymeasure.instruments import Instrument, Channel, validators
+from pyvisa.constants import Parity, StopBits
+
+import pyvisa as pv
+from pyvisa import VisaIOError  # noqa: F401 for dependencies
+
+
+def calculateChecksum(message):
+    """Calculate the checksum for string `message`."""
+    value = 0
+    for i in range(len(message)):
+        value += ord(message[i])
+    return chr(value % 64 + 64)
+
+
+def compose_data(value):
+    """Generate a string with the length of `value` and the `value` itself afterwards.
+
+    :param value: Value to send to the device.
+    """
+    value = f"{value}"
+    return f"{len(value):02}{value}"
+
+
+source = {
+    'combination': 0,
+    'pirani': 1,
+    'piezo': 2,
+    'hotCathode': 3,
+    'coldCathode': 4,
+    'ambient': 6,
+    'relative': 7,
+}
+
+
+class Sources(IntEnum):
+    COMBINATION = 0
+    PIRANI = 1
+    PIEZO = 2
+    HOT_CATHODE = 3
+    COLD_CATHODE = 4
+    AMBIENT = 6
+    RELATIVE = 7
+
+
+gas_factor = Channel.control(
+    "0C{ch}00", "0C{ch}%s", "Control the gas correction factor.",
+    values=(0.2, 8),
+    validator=validators.strict_range,
+    set_process=compose_data,
+)
+
+
+class SensorChannel(Channel):
+    """Generic channel for individual pressure sensors of a Transmitter."""
+    _id = -1  # obligatory channel number, define in channel types
+
+    def __init__(self, parent, id):
+        if id is None:
+            id = self._id
+        assert id == self._id, f"Pirani ID has to be {self._id}."
+        super().__init__(parent, id=self._id)
+
+    pressure = Channel.measurement(
+        "0M{ch}00", """Get the current pressure in mbar.""",
+        preprocess_reply=lambda r: r.replace("UR", "0").replace("OR", "inf"),
+    )
+
+
+class Pirani(SensorChannel):
+    """Pirani sensor channel.
+
+    A Pirani sensor measures the heat transfer in the gas. The reading depends on the gas type.
+    The sensor reading is linear to the real pressure below some threshold, around 1 mbar.
+    You may control a gas factor with :attr:`gas_factor`.
+    """
+
+    _id = 1
+
+    gas_factor = gas_factor
+
+    statistics = Channel.measurement(
+        "0PM011",
+        """Get the sensor statistics: wear in percent (negative: corrosion,
+            positive: contamination), time since last adjustment in hours.""",
+        separator="A",
+        cast=int,
+        get_process=lambda vals: [v / 4 for v in vals],
```

Comment:
Let me add that my device seems to support this command and that here really the processing of the reply seems buggy. As far as I see from the response (consistent with the documentation) there are some letters in the reply which separate the different values which somehow will have to be cut away.

Reply:
Please give me the response of the device, such that we can use it to write a test.

Reply:
```
import pymeasure.instruments
vsh = pymeasure.instruments.thyracont.VSH("...")
vsh.write("0PM011")  # message transmitted "0010PM011p"
vsh.read()  # raw return b'0011PM06W0A382j'
# returns 'W0A382'
```
This corresponds to ~95 hours since zero adjust and 0% wear.

Reply:
```suggestion
        preprocess_reply=lambda msg: msg.strip("W"),
        get_process=lambda vals: [vals[0], vals[1]/4],
```

This would work for me...

Reply:
Thanks for that communication example. I just verified it with my device.

Reply:
done.